src/librustc_parse/parser/mod.rs

```diff
@@ -184,9 +179,22 @@ impl TokenCursor {
                 return Token::new(token::Eof, DUMMY_SP);
             };
 
-            match self.frame.last_token {
-                LastToken::Collecting(ref mut v) => v.push(tree.clone().into()),
-                LastToken::Was(ref mut t) => *t = Some(tree.clone().into()),
+            // Don't set an open delimiter as our current token - we want
+            // to leave it as the full `TokenTree::Delimited` from the previous
+            // iteration of this loop
+            if !matches!(tree, TokenTree::Token(Token { kind: TokenKind::OpenDelim(_), .. })) {
+                self.cur_token = Some(tree.clone().into());
+            }
+
+            if let Some(collecting) = &mut self.collecting {
+                if collecting.depth == self.stack.len() {
+                    debug!(
+                        "TokenCursor::next():  collected {:?} at depth {:?}",
+                        tree,
+                        self.stack.len()
+                    );
+                    collecting.buf.push(tree.clone().into())
```

Comment:
Looks like we are losing jointness when collecting tokens.
This is pre-existing, but should be easy to fix with
```diff
-} else if let Some(tree) = self.frame.tree_cursor.next() {
+} else if let Some(tree) = self.frame.tree_cursor.next_with_joint() {
```
above.